stout/iterate.h

```diff
@@ -0,0 +1,136 @@
+#include <array>
+#include <iterator>
+
+#include "stout/context.h"
+#include "stout/loop.h"
+#include "stout/stream.h"
+
+////////////////////////////////////////////////////////////////////////
+
+namespace stout {
+namespace eventuals {
+
+////////////////////////////////////////////////////////////////////////
+
+template <typename Iter>
+auto Iterate(Iter begin, Iter end) {
+  struct ContainerData {
+    Iter it_;
+    Iter end_;
+  };
+
+  using T = typename std::iterator_traits<Iter>::value_type;
+
+  return Stream<T>()
+      .context(ContainerData{begin, end})
+      .next([](auto& containerData, auto& k) {
+        if (containerData.it_ != containerData.end_) {
+          k.Emit(*(containerData.it_++));
+        } else {
+          k.Ended();
+        }
+      })
+      .done([](auto&, auto& k) {
+        k.Ended();
+      });
+}
+
+template <typename Container>
+auto Iterate(Container& container) {
+  return Iterate(container.cbegin(), container.cend());
+}
+
+template <typename Container>
+auto Iterate(Container&& container) {
+  using CIter = typename Container::const_iterator;
+
+  struct ContainerData {
+    CIter it_;
+    Container data_;
+  };
+
+  using T = typename Container::value_type;
+
+  return Stream<T>()
+      .context(ContainerData{container.cbegin(), std::move(container)})
```

Comment:
This seems like it could be a bug, you're getting the beginning of a container but then moving that container? Are iterators valid even after the container has been moved? Seems like we want to store an `std::optional` of the type returned from `container.cbegin()` and then initialize it in the `.start()` callback. In that case we will need a `struct Data`, but let's call the fields `begin` and `container` to match the names used in the function scope.

Also your code will have to change slightly because you'll need to dereference `begin` since it'll be in a `std::optional`.